File: CourseTech.Application/Services/UserProfileService.cs
using CourseTech.Application.Commands.UserProfileCommands;
using CourseTech.Domain;
using CourseTech.Domain.Constants.Cache;
using CourseTech.Domain.Dto.UserProfile;
using CourseTech.Domain.Interfaces.Cache;
using CourseTech.Domain.Interfaces.Repositories;
using CourseTech.Domain.Interfaces.Services;
using MediatR;

namespace CourseTech.Application.Services;

public class UserProfileService(
    ICacheService cacheService,
    IUserProfileRepository userProfileRepository,
    IMediator mediator) : IUserProfileService
{
    /// <inheritdoc/>
    public async Task<Result<UserProfileDto>> GetUserProfileAsync(Guid userId)
    {
        var profileDto = await cacheService.GetOrAddToCache(
            $"{CacheKeys.UserProfile}{userId}",
            async () => await userProfileRepository.GetByUserIdAsync(userId));

        if (profileDto is null)
        {
            return Result<UserProfileDto>.Failure("User profile not found");
        }

        return Result.Success(new UserProfileDto()); //TODO заглушечка впадлу мапить пока
    }

    /// <inheritdoc/>
    public async Task<Result> UpdateUserProfileAsync(UpdateUserProfileDto dto, Guid userId)
    {
        var profile = await userProfileRepository.GetByUserIdAsync(userId);
        if (profile is null)
        {
            return Result.Failure("User profile not found");
        }

        await mediator.Send(new UpdateUserProfileCommand(dto, profile));

        var userProfileKey = $"{CacheKeys.UserProfile}:{userId}";

        await cacheService.RemoveAsync(userProfileKey);

        await cacheService.SetObjectAsync(userProfileKey, profile);

        return Result.Success();
    }
}

File: coursetech.client/src/services/user-profile-service.ts
import { RouteConstants } from "../constants/route-constants";
import { UpdateUserProfileDto } from "../types/dto/userProfile/update-user-profile-dto";
import { UserProfileDto } from "../types/dto/userProfile/user-profile-dto";
import { BaseResult } from "../types/result/base-result";
import { DataResult } from "../types/result/data-result";
import { ApiClient } from "./api-client";

export class UserProfileService {
    private apiClient: ApiClient;

    constructor(baseUrl: string) {
        this.apiClient = new ApiClient(baseUrl);
    }

    public async getUserProfileAsync(userId : string): Promise<DataResult<UserProfileDto>> {
        return this.apiClient.get<DataResult<UserProfileDto>>(RouteConstants.GET_USER_PROFILE, { userId: userId });
    }

    public async updateUserProfileAsync(dto: UpdateUserProfileDto): Promise<BaseResult> {
        return this.apiClient.put<BaseResult>(RouteConstants.UPDATE_USER_PROFILE, dto);
    }
}
